presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergFilterPushdown.java

```diff
@@ -0,0 +1,287 @@
+
+            this.resourceFactory = requireNonNull(resourceFactory, "resourceFactory is null");
+            this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+            this.typeManager = requireNonNull(typeManager, "typeManager is null");
+            this.icebergTransactionManager = requireNonNull(icebergTransactionManager, "transactionManager is null");
+        }
+
+        @Override
+        protected ConnectorPushdownFilterResult getConnectorPushdownFilterResult(
+                Map<String, ColumnHandle> columnHandles,
+                ConnectorMetadata metadata,
+                ConnectorSession session,
+                RemainingExpressions remainingExpressions,
+                DomainTranslator.ExtractionResult<Subfield> decomposedFilter,
+                RowExpression optimizedRemainingExpression,
+                Constraint<ColumnHandle> constraint,
+                Optional<ConnectorTableLayoutHandle> currentLayoutHandle,
+                ConnectorTableHandle tableHandle)
+        {
+            Table icebergTable;
+            if (metadata instanceof IcebergHiveMetadata) {
+                ExtendedHiveMetastore metastore = ((IcebergHiveMetadata) metadata).getMetastore();
+                icebergTable = getHiveIcebergTable(metastore, hdfsEnvironment, session, ((IcebergTableHandle) tableHandle).getSchemaTableName());
+            }
+            else {
+                icebergTable = getNativeIcebergTable(resourceFactory, session, ((IcebergTableHandle) tableHandle).getSchemaTableName());
+            }
+
+            TupleDomain<ColumnHandle> unenforcedConstraint = TupleDomain.withColumnDomains(Maps.filterKeys(constraint.getSummary().getDomains().get(), not(Predicates.in(getPartitionKeyColumnHandles(icebergTable, typeManager)))));
+
+            TupleDomain<Subfield> domainPredicate = getDomainPredicate(decomposedFilter, unenforcedConstraint);
+
+            Set<String> predicateColumnNames = getPredicateColumnNames(optimizedRemainingExpression, domainPredicate);
+
+            Map<String, IcebergColumnHandle> predicateColumns = predicateColumnNames.stream()
+                    .map(columnHandles::get)
+                    .map(IcebergColumnHandle.class::cast)
+                    .collect(toImmutableMap(IcebergColumnHandle::getName, Functions.identity()));
+
+            Optional<Set<IcebergColumnHandle>> requestedColumns = currentLayoutHandle.map(layout -> ((IcebergTableLayoutHandle) layout).getRequestedColumns()).orElse(Optional.empty());
+            TupleDomain<ColumnHandle> partitionColumnPredicate = TupleDomain.withColumnDomains(Maps.filterKeys(constraint.getSummary().getDomains().get(), Predicates.in(getPartitionKeyColumnHandles(icebergTable, typeManager))));
+
+            List<IcebergColumnHandle> partitionColumns = getPartitionKeyColumnHandles(icebergTable, typeManager);
```

Comment:
getPartitionKeyColumnHandles() are called twice but they should return the same result. This function goes over all fields in O(n) and not good for table with many columns. Would it be ok to move line 197 before 195 and just use partitionColumns there?